subprojects/docs/src/docs/userguide/java.adoc

```diff
@@ -0,0 +1,810 @@
+
+`forkEvery` - default: 0 (no max)::
+This property specifies the maximum number of test classes that Gradle should run on a test process before its disposed of and a fresh one created. This is mainly used as a way to manage memory consumption during the test execution phase of the build.
+
+`ignoreFailures` — default: false::
+If this property is `true`, Gradle will continue with the project's build once the tests have completed, even if some of them have failed. Note that, by default, the `Test` task always executes every test that it detects, irrespective of this setting.
+
+`failFast` —  (since Gradle 4.6) default: false::
+Set this to `true` if you want the build to fail and finish as soon as one of your tests fails. This can save a lot of time when you have a long-running test suite and is particularly useful when running the build on continuous integration servers. When a build fails before all tests have run, the test reports only include the results of the tests that have completed, successfully or not.
++
+You can also enable this behavior by using the `--fail-fast` command line option.
+
+`testLogging` — default: _not set_::
+This property represents a set of options that control which test events are logged and at what level. You can also configure other logging behavior via this property. See api:org.gradle.api.tasks.testing.logging.TestLoggingContainer[] for more detail.
+
+See api:org.gradle.api.tasks.testing.Test[] for details on all the available configuration options.
+[NOTE]
+====
+
+The test process can exit unexpectedly if configured incorrectly. For instance, if the Java executable does not exist or an invalid JVM argument is provided, the test process will fail to start. Similarly, if a test makes programmatic changes to the test process, this can also cause unexpected failures.
+
+For example, issues may occur if a `{javaApi}/java/lang/SecurityManager.html[SecurityManager]` is modified in a test because
+Gradle's internal messaging depends on reflection and socket communication, which may be disrupted if the permissions on the security manager change. In this particular case, you should restore the original `SecurityManager` after the test so that the
+gradle test worker process can continue to function.
+
+====
+
+[[sec:debugging_java_tests]]
+==== Debugging
+
+On the few occasions that you want to debug your code while the tests are running, it can be helpful if you can attach a debugger at that point. You can either set the api:org.gradle.api.tasks.testing.Test#getDebug()[] property to `true` or use the `--debug-jvm` command line option.
+
+When debugging for tests is enabled, Gradle will start the test process suspended and listening on port 5005.
+
+[[test_filtering]]
+==== Test filtering
```

Comment:
There's a lot of content on testing. We maybe want to split it out into its own chapter.

Reply:
Yup. I think Eric's OK with that too.

- [ ] Refactor testing into its own chapter